lib/ash_authentication/audit_log_resource/batcher.ex

```diff
@@ -0,0 +1,191 @@
+          timeout: Info.audit_log_write_batching_timeout!(resource),
+          max_size: Info.audit_log_write_batching_max_size!(resource),
+          action: Info.audit_log_write_action_name!(resource),
+          domain: Info.audit_log_domain!(resource),
+          queue: [],
+          queue_size: 0,
+          flush_timer: nil
+        })
+      end)
+
+    if map_size(audit_configs) == 0 do
+      :ignore
+    else
+      {:ok, audit_configs}
+    end
+  end
+
+  @doc false
+  @impl true
+  def handle_info({:write_batch, resource}, state) do
+    state =
+      state
+      |> Map.replace_lazy(resource, &write_batch(resource, %{&1 | flush_timer: nil}))
+
+    {:noreply, state}
+  end
+
+  @doc false
+  @impl true
+  def handle_cast({:enqueue, changeset}, state) do
+    state =
+      state
+      |> enqueue(changeset)
+      |> maybe_write_batches()
+
+    {:noreply, state}
+  end
+
+  @doc false
+  @impl true
+  def handle_call(:flush, _from, state) do
+    state =
+      state
+      |> Enum.reduce(%{}, fn {resource, config}, configs ->
+        config =
+          if config.queue_size > 0 do
+            if is_reference(config.flush_timer) do
+              Process.cancel_timer(config.flush_timer)
+            end
+
+            write_batch(resource, %{config | flush_timer: nil})
+          else
+            config
+          end
+
+        Map.put(configs, resource, config)
+      end)
+
+    {:reply, :ok, state}
+  end
+
+  @doc false
+  @impl true
+  def terminate(_reason, state) do
+    state
+    |> Enum.filter(fn {_, %{queue_size: qs}} -> qs > 0 end)
+    |> Enum.each(fn {resource, config} ->
+      write_batch(resource, config)
+    end)
+  end
+
+  defp enqueue(state, changeset) do
+    Map.update!(state, changeset.resource, fn
+      config when config.queue_size == 0 and config.queue == [] and config.timeout >= 0 ->
+        timer = Process.send_after(self(), {:write_batch, changeset.resource}, config.timeout)
+
+        %{
+          config
+          | queue: [changeset],
+            queue_size: 1,
+            flush_timer: timer
+        }
+
+      config ->
+        %{config | queue: [changeset | config.queue], queue_size: config.queue_size + 1}
+    end)
+  end
+
+  defp maybe_write_batches(state) do
+    state
+    |> Enum.reduce(%{}, fn
+      {resource, config}, configs when config.queue_size >= config.max_size ->
+        config =
+          if is_reference(config.flush_timer) do
+            Process.cancel_timer(config.flush_timer)
+            write_batch(resource, %{config | flush_timer: nil})
+          else
+            write_batch(resource, config)
+          end
+
+        Map.put(configs, resource, config)
+
+      {resource, config}, configs ->
+        Map.put(configs, resource, config)
+    end)
+  end
+
+  defp write_batch(resource, config) do
+    bulk_result =
+      config.queue
+      |> Stream.map(& &1.attributes)
+      |> Ash.bulk_create(resource, config.action,
+        authorize?: false,
```

Comment:
should we really do this? shouldn't this be handled by the policy that checks the context?

Reply:
But no policies are added by the igniter/transformer at the moment, maybe we should add it there and remvoe the authorize?: false here.

Reply:
Yeah, this definitely should use the ash_authentication context key.